### PR TITLE
Added significantly more early null checks throughout Spotless.

### PR DESCRIPTION
**Pull request by jbduncan**
Issue #86.

### CODE REVIEW

**Comment by jbduncan** [comment]
@nedtwigg I've just found the time to finish inserting the null checks I proposed in issue #86.

May I have your thoughts on the changes I've made?

**Comment by nedtwigg** [comment]
1. Moving the tests into lib breaks the eclipse projects.

lib_test -> testlib -> lib

But eclipse doesn't make a distinction between lib and lib_test, so it's a circular dependency.  That's why the tests have lived in testlib rather than lib/test.

2. Not a deal-breaker, but ideally pure-formatting changes would be in their own commit.  `Polishing` is a combination of formatting changes and impl changes.

3. The null-checking looks good.

**Comment by jbduncan** [comment]
> 1. Moving the tests into lib breaks the eclipse projects.
> lib_test -> testlib -> lib
>
> But eclipse doesn't make a distinction between lib and lib_test, so it's a circular dependency. That's why the tests have lived in testlib rather than lib/test.

Oh that's a shame!

I'll put the lib tests back in testlib then, as I originally did it to better allow me to test my changes with `NullPointerTester` (which I eventually gave up using).

> 2. Not a deal-breaker, but ideally pure-formatting changes would be in their own commit. `Polishing` is a combination of formatting changes and impl changes. Formatting changes don't need much review, but impl does. Mixing them up means the whole commit needs review, even though most of it is non-content.

Okey dokey, I'll look into splitting that commit up.

**Comment by jbduncan** [comment]
@nedtwigg I've made some new changes, whilst breaking up changes from the last set of commits (now gone due to force-push) into more logical ones.

Would you kindly have another look?

**Comment by jbduncan** [comment]
Wow, thank you for such a thorough analysis @nedtwigg!

> Very minor spot - sometimes `requireNonNull` has a static import, othertimes not. Doesn't bother me either way, but this is a very "self-consistent" batch of commits, so figured I'd point it out in case it bothers you :)

I usually would be bothered by this (good observation!), but in this case I don't think it's worth the time and effort, so I'll happily leave them as-is. :)

> I'd like to zoom out a little bit and do some cost-benefit on this change...
> 
> ...So until there's an automated rule that enforces [null checks], I don't want to enforce this style on future contributions. :)

I agree with everything you've said here!

We could potentially use the Checker Framework to do some sort of auto-null-analysis in the future, but I doubt it's worth the effort for a project as small as Spotless.

> ...I think the @Nullable in FormatExceptionPolicy is wrong.

Oh yes, you're right!

I made it `@Nullable` because the (only?) place where `FormatExceptionPolicy.handleError()` is used is in `Formatter.compute()` (which is used to handle any `Throwable`s that get thrown by any executed `FormatterStep`), and I thought that it was possible to throw `null`s in Java. However, I've been proved mistaken according to this [StackOverflow post](http://stackoverflow.com/questions/17576922/why-can-i-throw-null-in-java).

I'll change it as soon as I've the time.

File: lib/src/main/java/com/diffplug/spotless/FormatExceptionPolicy.java [hunk 1]
```diff
@@ -17,10 +17,12 @@
 
 import java.io.Serializable;
 
+import javax.annotation.Nullable;
+
 /** A policy for handling exceptions in the format. */
 public interface FormatExceptionPolicy extends Serializable, NoLambda {
 	/** Called for every error in the formatter. */
-	void handleError(Throwable e, FormatterStep step, String relativePath);
+	public void handleError(@Nullable Throwable e, FormatterStep step, String relativePath);
```
Comment (jbduncan) [inline lib/src/main/java/com/diffplug/spotless/FormatExceptionPolicy.java]: See the bottom of my response [here](https://github.com/diffplug/spotless/pull/101#issuecomment-294349655). :)

**Comment by nedtwigg** [comment]
Just a ping that this PR is at high risk for causing merge conflicts.

**Comment by jbduncan** [comment]
Thanks for the warning @nedtwigg. I've put this higher up my to-do list. :)

**Comment by jbduncan** [comment]
@nedtwigg What do you think now?

**Comment by nedtwigg** [comment]
Needs a `spotlessApply` ;-)

**Comment by jbduncan** [comment]
Ta. :+1:

**Pull request merged**